pkg/controller/security/manager.go

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright 2024 The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package security
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"container/heap"
+
+	istiosecurity "istio.io/istio/pkg/security"
+
+	"kmesh.net/kmesh/pkg/constants"
+	"kmesh.net/kmesh/pkg/logger"
+)
+
+type certExp struct {
+	Identity string
+	exp      time.Time
+	index    int
+}
+
+type certsCache struct {
+	cert   *istiosecurity.SecretItem
+	refCnt int32
+}
+
+type SecretManagerCache struct {
+	caClient *CaClient
+
+	// configOptions includes all configurable params for the cache.
+	configOptions *istiosecurity.Options
+
+	// storing certificates
+	certsCache *sync.Map
+
+	//caRootPath string
+	pending *PriorityQueue
+}
+
+var log = logger.NewLoggerField("security")
+
+type PriorityQueue struct {
+	queue []*certExp
+	mu    sync.Mutex
+}
+
+func NewPriorityQueue() *PriorityQueue {
+	return &PriorityQueue{
+		queue: make([]*certExp, 0),
+		mu:    sync.Mutex{},
+	}
+}
+
+func (pq *PriorityQueue) Push(x interface{}) {
+	n := len(pq.queue)
+	item := x.(*certExp)
+	item.index = n
+	pq.queue = append(pq.queue, item)
+}
+
+func (pq *PriorityQueue) Pop() interface{} {
+	old := pq.queue
+	n := len(old)
+	x := old[n-1]
+	old[n-1] = nil // avoid memory leak
+	x.index = -1
+	pq.queue = old[0 : n-1]
+	return x
+}
+
+func (pq *PriorityQueue) Len() int {
+	return len(pq.queue)
+}
+
+func (pq *PriorityQueue) Less(i, j int) bool {
+	return pq.queue[i].exp.Before(pq.queue[j].exp)
+}
+
+func (pq *PriorityQueue) Swap(i, j int) {
+	pq.queue[i], pq.queue[j] = pq.queue[j], pq.queue[i]
+}
+
+func (pq *PriorityQueue) addItem(certExp *certExp) {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	heap.Push(pq, certExp)
+}
+
+func (pq *PriorityQueue) delete(Identity string) *certExp {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	for i := 0; i < len(pq.queue); i++ {
+		log.Debugf("delete item.Identity:%v\n", pq.queue[i].Identity)
+		if pq.queue[i].Identity == Identity {
+			return heap.Remove(pq, i).(*certExp)
+		}
+	}
+	return nil
+}
+
+func (pq *PriorityQueue) lookTop() *certExp {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	return pq.queue[0]
+}
+
+// NewsecretManager creates a new secretManager.
+func NewsecretManager() (*SecretManagerCache, error) {
```

Comment:
Usually we make the return type align with the construction name
```suggestion
func NewSecretManager() (*SecretManager, error) {
```

Reply:
ok